src/utilities/ntrace/PluginLogWriter.cpp

```diff
@@ -132,23 +158,35 @@ FB_SIZE_T PluginLogWriter::write(const void* buf, FB_SIZE_T size)
 
 	if (m_maxSize && (fileSize > m_maxSize))
 	{
-		const TimeStamp stamp(TimeStamp::getCurrentTimeStamp());
-		struct tm times;
-		stamp.decode(&times);
-
 		PathName newName;
-		const FB_SIZE_T last_dot_pos = m_fileName.rfind(".");
-		if (last_dot_pos > 0)
-		{
-			PathName log_name = m_fileName.substr(0, last_dot_pos);
-			PathName log_ext = m_fileName.substr(last_dot_pos + 1, m_fileName.length());
-			newName.printf("%s.%04d-%02d-%02dT%02d-%02d-%02d.%s", log_name.c_str(), times.tm_year + 1900,
-				times.tm_mon + 1, times.tm_mday, times.tm_hour, times.tm_min, times.tm_sec, log_ext.c_str());
-		}
-		else
+
+		while (true)
 		{
-			newName.printf("%s.%04d-%02d-%02dT%02d-%02d-%02d", m_fileName.c_str(), times.tm_year + 1900,
-				times.tm_mon + 1, times.tm_mday, times.tm_hour, times.tm_min, times.tm_sec);
+			const TimeStamp stamp(TimeStamp::getCurrentTimeStamp());
+			struct tm times;
+			int fractions;
+			stamp.decode(&times, &fractions);
+
+			const FB_SIZE_T last_dot_pos = m_fileName.rfind(".");
+			if (last_dot_pos > 0)
+			{
+				PathName log_name = m_fileName.substr(0, last_dot_pos);
+				PathName log_ext = m_fileName.substr(last_dot_pos + 1, m_fileName.length());
+				newName.printf("%s.%04d-%02d-%02dT%02d-%02d-%02d-%04d.%s", log_name.c_str(), times.tm_year + 1900,
+					times.tm_mon + 1, times.tm_mday, times.tm_hour, times.tm_min, times.tm_sec, fractions, log_ext.c_str());
+			}
+			else
+			{
+				newName.printf("%s.%04d-%02d-%02dT%02d-%02d-%02d-%04d", m_fileName.c_str(), times.tm_year + 1900,
+					times.tm_mon + 1, times.tm_mday, times.tm_hour, times.tm_min, times.tm_sec, fractions);
+			}
+
+			struct stat st;
+
+			if (stat(newName.c_str(), &st))
+				break;
```

Comment:
> Why generate another name in a loop - you should first reopen current log and check its size again, isn't is ?
If two threads generate same name (don't tell me it is impossible), one of them will rename current log full of records, then second one will generate another name (few ms later) and rename current (almost empty) log.

I didn't get it. While the first thread is generating the name and renaming the current file, the second one is waiting in the beginning of `write()` function. When the first one returns from `write()`, the second one will continue, check the size of the current file (it's already renamed at the moment), reopen it (because the size > `m_maxSize`) and check the size again to decide whether to rename the file or just write to it.

> Also, taking into account that code is now protected by mutex, I see no need to add fractions into log file name.

There is a case when rotation is performed more often than once per second, and the server may hang until the end of a second. I know that such configurations are wrong but it's still possible. If the name format is important for some reason, I can remove fractions.

Reply:
@aafemt, the mutex is shared.

Reply:
Yes, it works the same way in classic mode.